File: quickshell/net-tracker.py
#!/usr/bin/env python3
import os, sys
# Delegate immediately to high-performance native Rust binary if present
_rust_bin = os.path.join(os.path.dirname(os.path.abspath(__file__)), "net-tracker")
if os.path.isfile(_rust_bin) and os.access(_rust_bin, os.X_OK):
    os.execv(_rust_bin, [_rust_bin] + sys.argv[1:])

import sys
import os
import glob
import json
import time
import subprocess

DATA_DIR = os.path.expanduser("~/.local/share/quickshell")
USAGE_FILE = os.path.join(DATA_DIR, "network_usage.json")

def get_boot_id():
    try:
        with open("/proc/sys/kernel/random/boot_id", "r") as f:
            return f.read().strip()
    except Exception:
        return "unknown"

def fmt_bytes(b):
    if b >= 1024**3:
        # Show 3 decimal places for GB (e.g., 1.234 GB)
        return f"{b / (1024**3):.3f} GB"
    elif b >= 1024**2:
        return f"{b / (1024**2):.1f} MB"
    elif b >= 1024:
        return f"{b / 1024:.0f} KB"
    else:
        return f"{b} B"

def is_physical_interface(dev_path):
    iface = os.path.basename(dev_path)
    # Exclude virtual / container / VPN tunnels
    if iface.startswith(("lo", "docker", "veth", "virbr", "br-", "tun", "wg", "tailscale", "dummy", "zt")):
        return False
    # Check if sysfs reports a hardware device link
    if os.path.exists(os.path.join(dev_path, "device")):
        return True
    # Fallback to physical naming schemes (Wi-Fi, Ethernet, USB tethering, WWAN)
    if iface.startswith(("wlan", "wlp", "enp", "eth", "eno", "usb", "enx", "wwan", "wwp")):
        return True
    return False

def update_usage():
    os.makedirs(DATA_DIR, exist_ok=True)
    cur_date = time.strftime("%Y-%m-%d")
    cur_month = time.strftime("%Y-%m")
    boot_id = get_boot_id()

    data = {}
    if os.path.exists(USAGE_FILE):
        try:
            with open(USAGE_FILE, "r") as f:
                data = json.load(f)
        except Exception:
            data = {}

    saved_boot_id = data.get("boot_id", "")
    saved_date = data.get("date", cur_date)
    saved_month = data.get("month", cur_month)
    day_bytes = data.get("day_bytes", 0)
    month_bytes = data.get("month_bytes", 0)
    saved_ifaces = data.get("ifaces", {})

    # Rollover logic
    if cur_date != saved_date:
        day_bytes = 0
        saved_date = cur_date
    if cur_month != saved_month:
        month_bytes = 0
        saved_month = cur_month

    is_new_boot = (boot_id != saved_boot_id)

    total_delta = 0

    for dev_path in glob.glob("/sys/class/net/*"):
        if not is_physical_interface(dev_path):
            continue
        iface = os.path.basename(dev_path)
        try:
            with open(os.path.join(dev_path, "statistics/rx_bytes")) as f:
                rx = int(f.read().strip())
            with open(os.path.join(dev_path, "statistics/tx_bytes")) as f:
                tx = int(f.read().strip())
            total = rx + tx

            if iface not in saved_ifaces:
                # Newly detected interface: set baseline without adding past uptime
                saved_ifaces[iface] = total
                delta = 0
            elif is_new_boot:
                # System rebooted: kernel counter started from 0
                delta = total
                saved_ifaces[iface] = total
            else:
                prev = saved_ifaces[iface]
                if total >= prev:
                    delta = total - prev
                else:
                    # Interface reconnected or counter wrapped
                    delta = total
                saved_ifaces[iface] = total

            total_delta += delta
        except Exception:
            pass

    day_bytes += total_delta
    month_bytes += total_delta

    data = {
        "boot_id": boot_id,
        "date": cur_date,
        "month": cur_month,
        "day_bytes": day_bytes,
        "month_bytes": month_bytes,
        "ifaces": saved_ifaces
    }

    try:
        with open(USAGE_FILE + ".tmp", "w") as f:
            json.dump(data, f, indent=2)
        os.replace(USAGE_FILE + ".tmp", USAGE_FILE)
    except Exception:
        pass

    # Trigger daily markdown logger update
    try:
        script_dir = os.path.dirname(os.path.abspath(__file__))
        logger_script = os.path.join(script_dir, "daily-network-logger.py")
        if os.path.exists(logger_script):
            subprocess.Popen(["python3", logger_script, "--update"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    except Exception:
        pass

    return day_bytes, month_bytes

def get_active_network():
    try:
        proc = subprocess.run(
            ["nmcli", "-t", "-f", "TYPE,STATE,CONNECTION,DEVICE", "dev"],
            capture_output=True, text=True, timeout=2
        )
        for line in proc.stdout.strip().split("\n"):
            if not line:
                continue
            parts = line.split(":")
            if len(parts) >= 3 and "connected" in parts[1]:
                t, conn, dev = parts[0].lower(), parts[2], parts[3] if len(parts) > 3 else ""
                if dev == "lo" or "loopback" in t:
                    continue
                if t == "wifi":
                    return "wifi", conn
                elif t == "ethernet":
                    if "usb" in dev.lower() or "enx" in dev.lower() or "rndis" in dev.lower() or "usb" in conn.lower():
                        return "usb", (conn if conn else "USB Tethering")
                    else:
                        return "ethernet", (conn if conn else "Ethernet")
                elif t in ["gsm", "cdma", "wwan"]:
                    return "cellular", (conn if conn else "Cellular")
        return "none", "Disconnected"
    except Exception:
        return "none", "Disconnected"

if __name__ == "__main__":
    action = sys.argv[1] if len(sys.argv) > 1 else "--get"
    if action == "--update":
        update_usage()
    elif action == "--type":
        net_type, net_name = get_active_network()
        print(f"{net_type}|{net_name}")
    else:
        d_bytes, m_bytes = update_usage()
        print(f"{fmt_bytes(d_bytes)}|{fmt_bytes(m_bytes)}")

File: dark-tools-rs/src/bin/net_tracker.rs
#[path = "../common.rs"]
mod common;

use chrono::Local;
use common::{fmt_bytes, get_boot_id, is_physical_interface};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::PathBuf;
use std::process::Command;

#[derive(Serialize, Deserialize, Debug, Default)]
struct UsageData {
    #[serde(default)]
    boot_id: String,
    #[serde(default)]
    date: String,
    #[serde(default)]
    month: String,
    #[serde(default)]
    day_bytes: u64,
    #[serde(default)]
    month_bytes: u64,
    #[serde(default)]
    ifaces: HashMap<String, u64>,
}

fn get_data_dir() -> PathBuf {
    let home = env::var("HOME").unwrap_or_else(|_| "/home/aryan".to_string());
    PathBuf::from(home).join(".local/share/quickshell")
}

fn update_usage() -> (u64, u64) {
    let data_dir = get_data_dir();
    let _ = fs::create_dir_all(&data_dir);
    let usage_file = data_dir.join("network_usage.json");

    let now = Local::now();
    let cur_date = now.format("%Y-%m-%d").to_string();
    let cur_month = now.format("%Y-%m").to_string();
    let boot_id = get_boot_id();

    let mut data: UsageData = if usage_file.exists() {
        fs::read_to_string(&usage_file)
            .ok()
            .and_then(|s| serde_json::from_str(&s).ok())
            .unwrap_or_default()
    } else {
        UsageData::default()
    };

    let is_new_boot = data.boot_id != boot_id;

    // Rollover logic
    if data.date != cur_date {
        data.day_bytes = 0;
        data.date = cur_date.clone();
    }
    if data.month != cur_month {
        data.month_bytes = 0;
        data.month = cur_month.clone();
    }

    let mut total_delta: u64 = 0;

    if let Ok(entries) = fs::read_dir("/sys/class/net") {
        for entry in entries.flatten() {
            let path = entry.path();
            if !is_physical_interface(&path) {
                continue;
            }

            let iface = match path.file_name().and_then(|n| n.to_str()) {
                Some(name) => name.to_string(),
                None => continue,
            };

            let rx_bytes = fs::read_to_string(path.join("statistics/rx_bytes"))
                .ok()
                .and_then(|s| s.trim().parse::<u64>().ok())
                .unwrap_or(0);
            let tx_bytes = fs::read_to_string(path.join("statistics/tx_bytes"))
                .ok()
                .and_then(|s| s.trim().parse::<u64>().ok())
                .unwrap_or(0);

            let total = rx_bytes + tx_bytes;

            let delta = match data.ifaces.get(&iface) {
                None => {
                    // Newly detected interface: set baseline without adding past uptime
                    data.ifaces.insert(iface.clone(), total);
                    0
                }
                Some(&prev) => {
                    let d = if is_new_boot {
                        // System rebooted: kernel counter started from 0
                        total
                    } else if total >= prev {
                        total - prev
                    } else {
                        // Counter wrapped or interface reconnected
                        total
                    };
                    data.ifaces.insert(iface.clone(), total);
                    d
                }
            };

            total_delta += delta;
        }
    }

    data.boot_id = boot_id;
    data.day_bytes += total_delta;
    data.month_bytes += total_delta;

    // Atomic write
    let tmp_file = data_dir.join("network_usage.json.tmp");
    if let Ok(json_str) = serde_json::to_string_pretty(&data) {
        if fs::write(&tmp_file, json_str).is_ok() {
            let _ = fs::rename(&tmp_file, &usage_file);
        }
    }

    // Trigger daily logger in the background
    trigger_daily_logger();

    (data.day_bytes, data.month_bytes)
}

fn trigger_daily_logger() {
    let home = env::var("HOME").unwrap_or_else(|_| "/home/aryan".to_string());
    
    // Check candidate paths for daily-network-logger binary or script
    let candidates = [
        // 1. Rust compiled binary in quickshell directory
        PathBuf::from(&home).join("DARK_NIRI/quickshell/daily-network-logger"),
        // 2. Binary in current executable's directory
        env::current_exe()
            .ok()
            .and_then(|p| p.parent().map(|dir| dir.join("daily-network-logger")))
            .unwrap_or_else(|| PathBuf::from("daily-network-logger")),
    ];

    for candidate in &candidates {
        if candidate.exists() {
            let _ = Command::new(candidate)
                .arg("--update")
                .spawn();
            break;
        }
    }
}

fn get_active_network() -> (String, String) {
    let output = Command::new("nmcli")
        .args(["-t", "-f", "TYPE,STATE,CONNECTION,DEVICE", "dev"])
        .output();

    if let Ok(out) = output {
        if let Ok(text) = String::from_utf8(out.stdout) {
            for line in text.lines() {
                let line = line.trim();
                if line.is_empty() {
                    continue;
                }
                let parts: Vec<&str> = line.split(':').collect();
                if parts.len() >= 3 && parts[1].contains("connected") {
                    let t = parts[0].to_lowercase();
                    let conn = parts[2].trim();
                    let dev = if parts.len() > 3 { parts[3].trim() } else { "" };

                    if dev == "lo" || t.contains("loopback") {
                        continue;
                    }

                    if t == "wifi" {
                        return ("wifi".to_string(), conn.to_string());
                    } else if t == "ethernet" {
                        let dev_lower = dev.to_lowercase();
                        let conn_lower = conn.to_lowercase();
                        if dev_lower.contains("usb")
                            || dev_lower.contains("enx")
                            || dev_lower.contains("rndis")
                            || conn_lower.contains("usb")
                        {
                            let name = if conn.is_empty() { "USB Tethering" } else { conn };
                            return ("usb".to_string(), name.to_string());
                        } else {
                            let name = if conn.is_empty() { "Ethernet" } else { conn };
                            return ("ethernet".to_string(), name.to_string());
                        }
                    } else if ["gsm", "cdma", "wwan"].contains(&t.as_str()) {
                        let name = if conn.is_empty() { "Cellular" } else { conn };
                        return ("cellular".to_string(), name.to_string());
                    }
                }
            }
        }
    }

    ("none".to_string(), "Disconnected".to_string())
}

fn main() {
    let action = env::args().nth(1).unwrap_or_else(|| "--get".to_string());

    match action.as_str() {
        "--update" => {
            update_usage();
        }
        "--type" => {
            let (net_type, net_name) = get_active_network();
            println!("{}|{}", net_type, net_name);
        }
        _ => {
            let (day_bytes, month_bytes) = update_usage();
            println!("{}|{}", fmt_bytes(day_bytes), fmt_bytes(month_bytes));
        }
    }
}
